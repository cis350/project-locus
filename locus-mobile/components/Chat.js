/* eslint-disable max-len */
import React, { useState, useRef, useEffect } from 'react';
import {
  View, Text, StyleSheet, ScrollView, TouchableHighlight, TextInput, Image, Alert,
} from 'react-native';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import { getClubChat, sendMessage, updateNotifications } from '../modules/api';

// get messages from currentChat once backend is available
export default function Chat({ currentChat, backToAllChat, user }) {
  const [chatMessages, setChatMessages] = useState([]);
  const [message, setMessage] = useState('');
  const [content, setContent] = useState('');
  const scrollViewRef = useRef();

  // setup all messages and async calls to get all messages every 2 seconds
  useEffect(() => {
    async function fetchMessages() {
      setChatMessages((await getClubChat(currentChat)).jsonContent);
    }
    fetchMessages();
    const interval = setInterval(() => {
      fetchMessages();
    }, 2000);
    return () => clearInterval(interval);
  }, []);

  // handle send message
  const isValidUrl = (url) => {
    try {
      const f = new URL(url);
      if (f) {
        return true;
      }
    } catch (e) {
      return false;
    }
    return true;
  };

  const displayContent = (image) => {
    if (image === '' || !isValidUrl(image)) {
      return false;
    }
    return true;
  };

  const getTime = ((dateMilli) => {
    const date = new Date(dateMilli);
    const hour = date.getHours();
    let minute = date.getMinutes();
    if (minute < 10) minute = `0${minute}`;
    return `${hour}:${minute}`;
  });
  const getDate = ((dateMilli) => {
    const date = new Date(dateMilli);
    const year = date.getFullYear();
    const month = ((date.getMonth() + 1).toString()).slice(-2);
    const day = (date.getDate().toString()).slice(-2);
    return `${month}-${day}-${year}`;
  });

  // function that will send the message the user types, update for backend later
  async function handleSendMessage() {
    if (message === '' && content === '') return;
    if (/\S/.test(message)) {
      // TODO: pass date as milliseconds, update content to image content
      const response = await (sendMessage(currentChat, user.email, message, content, (new Date()).getTime()));
      if (response.status === 201) await updateNotifications(currentChat, user.email);
      else Alert.alert('Message send failed');
      setChatMessages((await getClubChat(currentChat)).jsonContent);
      setMessage('');
      setContent('');
    } else Alert.alert('Message Needs a Body');
  }

  // setup messages view for specific chat
  const displayMessages = [];
  for (let i = 0; i < chatMessages.length; i += 1) {
    // align right if you are the sender
    if (chatMessages[i].userEmail === user.email) {
      displayMessages.push(
        <View style={{ alignItems: 'flex-end' }} key={`message${i}`}>
          {displayContent(chatMessages[i].content)
            ? <Image source={{ uri: chatMessages[i].content }} style={styles.img} /> : <View /> }
          <View style={styles.textBubble}>
            <Text style={{ fontSize: 15 }}>{chatMessages[i].message}</Text>
          </View>
          <Text style={{ fontSize: 10 }}>{chatMessages[i].fullName}</Text>
          <Text style={{ fontSize: 10 }}>{getDate(chatMessages[i].timeStamp)}</Text>
          <Text style={{ fontSize: 10 }}>{getTime(chatMessages[i].timeStamp)}</Text>
        </View>,
      );
    } else { // align left if you are not the sender
      displayMessages.push(
        <View style={{ alignItems: 'flex-start' }} key={`message${i}`}>
          {displayContent(chatMessages[i].content)
            ? <Image source={{ uri: chatMessages[i].content }} style={styles.img} /> : <View /> }
          <View style={styles.textBubble}>
            <Text style={{ fontSize: 15 }}>{chatMessages[i].message}</Text>
          </View>
          <Text style={{ fontSize: 10 }}>{chatMessages[i].fullName}</Text>
          <Text style={{ fontSize: 10 }}>{getDate(chatMessages[i].timeStamp)}</Text>
          <Text style={{ fontSize: 10 }}>{getTime(chatMessages[i].timeStamp)}</Text>
        </View>,
      );
    }
  }
  displayMessages.push(<View style={{ marginBottom: 25 }} key="spacefiller" />);

  return (
    <KeyboardAwareScrollView scrollToEnd={{ animated: true }}>
      <View style={styles.container}>
        <TouchableHighlight style={styles.button} underlayColor="#33E86F" onPress={() => backToAllChat()}>
          <Text style={{ fontSize: 28, color: 'white' }}>All Chats</Text>
        </TouchableHighlight>
        <Text style={{ fontSize: 28 }}>Current Chat</Text>
        <ScrollView
          style={styles.messagesContainer}
          // https://stackoverflow.com/questions/29310553/is-it-possible-to-keep-a-scrollview-scrolled-to-the-bottom
          ref={scrollViewRef}
          onContentSizeChange={() => scrollViewRef.current.scrollToEnd({ animated: true })}
        >
          {displayMessages}
        </ScrollView>
        <View style={styles.inputContainer}>
          <TextInput
            style={styles.input}
            placeholder="Message"
            placeholderTextColor="grey"
            onChangeText={setMessage}
            onSubmitEditing={() => handleSendMessage()}
            value={message}
          />
          <TextInput
            style={styles.input}
            placeholder="Image Link"
            placeholderTextColor="grey"
            onChangeText={setContent}
            onSubmitEditing={() => handleSendMessage()}
            value={content}
          />
          <TouchableHighlight style={styles.sendButton} underlayColor="#33E86F" onPress={() => handleSendMessage()}>
            <Text style={{ fontSize: 28, color: 'white' }}>Send Message</Text>
          </TouchableHighlight>
        </View>
      </View>
    </KeyboardAwareScrollView>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    paddingTop: 25,
  },
  messagesContainer: {
    marginTop: 5,
    backgroundColor: '#B5E48C',
    height: 300,
    width: 350,
    borderRadius: 10,
    padding: 20,
  },
  textBubble: {
    backgroundColor: '#6A9B72',
    borderRadius: 10,
    padding: 5,
    maxWidth: 250,
    marginBottom: 5,
  },
  button: {
    backgroundColor: '#6A9B72',
    borderRadius: 10,
    paddingVertical: 10,
    width: 250,
    justifyContent: 'center',
    alignItems: 'center',
    marginHorizontal: 10,
    marginVertical: 10,
  },
  sendButton: {
    backgroundColor: '#52B69A',
    borderRadius: 10,
    paddingVertical: 10,
    width: 250,
    justifyContent: 'center',
    alignItems: 'center',
    marginHorizontal: 10,
    marginVertical: 10,
  },
  inputContainer: {
    alignItems: 'center',
  },
  input: {
    backgroundColor: 'white',
    borderColor: 'gray',
    width: 250,
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
    marginVertical: 5,
  },
  img: {
    width: 100,
    height: 100,
  },
});
