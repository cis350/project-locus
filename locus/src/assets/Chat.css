.chat-stack-scrollable {
  height: 500px;
  width: 1200px;
  overflow-y: auto;
  float: left;
  position: relative;
  margin-left: -5px;
}
/* 
.talktext {
  padding: 1em;
  text-align: left;
  line-height: 1.5em;
}

.talktext p {
  -webkit-margin-before: 0em;
  -webkit-margin-after: 0em;
}

.container {
  padding: 5% 5%;
}

.talk-bubble {
  margin: 40px;
  display: inline-block;
  position: relative;
  width: 200px;
  height: auto;
  background-color: lightyellow;
}
.border {
  border: 8px solid #666;
}
.round {
  border-radius: 30px;
  -webkit-border-radius: 30px;
  -moz-border-radius: 30px;
} */
