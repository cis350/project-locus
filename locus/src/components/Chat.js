import React, { useState, useRef } from 'react';
import { v4 as uuidv4 } from 'uuid';
import {
  Button, Form, Container, Row, Col, Stack,
} from 'react-bootstrap';
import '../assets/Clubs.css';
import {
  getUserFullName, getUserClubs, getClubChat, sendMessage,
} from '../modules/storage';
import '../assets/Chat.css';

function Chat({ userEmail }) {
  const [currentChat, changeChat] = useState([]);
  const currentClub = useRef('');
  const message = useRef('');
  // clubs user is in
  const userClubs = getUserClubs(userEmail);
  // const stackElement = document.getElementsByClassName('chat-stack-scrollable')[0];
  // stackElement.scrollTop = stackElement.scrollHeight;

  const switchToChat = (clubName) => {
    currentClub.current = clubName;
    changeChat(getClubChat(clubName));
  };

  const switchToClubs = () => {
    currentClub.current = '';
    changeChat([]);
  };

  const parseInput = (e) => {
    message.current = e.target.value;
  };

  const submitMessage = () => {
    if (/\S/.test(message.current)) {
      sendMessage(currentClub.current, userEmail, message.current, new Date(), uuidv4());
    }
    const updatedChat = getClubChat(currentClub.current);
    message.current = '';
    document.getElementById('input-text').value = '';
    // stackElement.scrollTop = stackElement.scrollHeight;
    changeChat(updatedChat);
  };

  const showAllClubs = (() => (
    <div>
      <Stack gap={20}>
        <Row>
          <h1 className="chat-header">
            Your Chats &nbsp;
          </h1>
        </Row>
        <div className="chat-table">
          {userClubs.map((clubName) => (
            <div className="club-item" key={clubName}>
              <Button className="club-button" onClick={() => switchToChat(clubName)}>
                <Row>
                  <Col className="d-flex justify-content-center">
                    {clubName}
                  </Col>
                </Row>
              </Button>
            </div>
          ))}
        </div>
      </Stack>
    </div>
  ));

  const showAChat = (() => (
    <Container>
      <Button onClick={switchToClubs}>Go back</Button>
      <Stack>
        <div className="chat-stack-scrollable">
          {currentChat.map((mess) => (
            <Row>
              <Col className="chat-item" key={mess[3]}>
                <div className="imessage">
                  <p className="from-them">{mess[1]}</p>
                </div>
                <p>
                  {getUserFullName(mess[0])}
                  :
                  {mess[2].toString()}
                </p>
              </Col>
            </Row>
          ))}
        </div>
      </Stack>
      <Form>
        <Form.Group className="mb-3" controlId="formMessage">
          <Form.Label>Input Message</Form.Label>
          <Form.Control id="input-text" type="message" placeholder="Send a message!" onChange={parseInput} />
        </Form.Group>
        <Button variant="primary" onClick={submitMessage}> Send </Button>
      </Form>
    </Container>
  ));

  return (
    <div>
      <Container>
        {(currentClub.current === '') ? showAllClubs() : showAChat() }
      </Container>
    </div>
  );
}

export default Chat;
