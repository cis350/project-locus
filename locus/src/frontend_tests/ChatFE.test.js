/**
* @jest-environment jsdom
*/
import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import Chat from '../components/Chat';

const dummyEmail = '16fangd@gmail.com';
const dummyChatClubsNames = ['PVG', 'PennLabs', 'WUFC', 'WITG'];

test('analytics snapshot test', () => {
  const component = renderer.create(<Chat />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
