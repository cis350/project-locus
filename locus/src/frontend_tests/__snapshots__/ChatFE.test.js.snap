// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat snapshot test 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="container"
    >
      <div>
        <div
          className="vstack gap-20"
        >
          <div
            className="row"
          >
            <h1
              className="chat-header"
            >
              Your Chats Â 
            </h1>
          </div>
          <div
            className="chat-table"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
